include/seastar/util/log.hh

```diff
@@ -211,14 +222,40 @@ public:
         return __builtin_expect(level <= _level.load(std::memory_order_relaxed), false) && !silent;
     }
 
+#if FMT_VERSION >= 80000
+    /// logs to desired level if enabled, otherwise we ignore the log line
+    ///
+    /// \param fmt - {fmt} style format string
+    /// \param args - args to print string
+    ///
+    template <typename... Args>
+    void log(log_level level, fmt::format_string<Args...> fmt, Args&&... args) noexcept {
+        if (is_enabled(level)) {
+            try {
+                lambda_log_writer writer([&] (internal::log_buf::inserter_iterator it) {
+                    return fmt::format_to(it, fmt, std::forward<Args>(args)...);
+                });
+                do_log(level, writer);
+            } catch (...) {
+                failed_to_log(std::current_exception(), current_backtrace());
+            }
+        }
+    }
+#endif
     /// logs to desired level if enabled, otherwise we ignore the log line
     ///
     /// \param fmt - {fmt} style format string (implictly converted to struct logger::format_info)
     ///              or a logger::format_info passed down the call chain.
     /// \param args - args to print string
     ///
-    template <typename... Args>
-    void log(log_level level, format_info fmt, Args&&... args) noexcept {
+    template <typename Fmt, typename... Args>
+#if FMT_VERSION >= 80000
+    [[deprecated("Use log(log_level, fmt::format_string<...>, ...) instead")]]
+    std::enable_if_t<std::is_same_v<Fmt, format_info>>
```

Comment:
Why is enable_if needed? Can't you just write format_info for the parameter's type?

Reply:
without templating this parameter, we'd confuse the compiler, and we'd have FTBFS like
```
/usr/bin/clang++ -DFMT_SHARED -DSEASTAR_API_LEVEL=7 -DSEASTAR_BUILD_SHARED_LIBS -DSEASTAR_DEBUG -DSEASTAR_DEBUG_SHARED_PTR -DSEASTAR_DEFAULT_ALLOCATOR -DSEASTAR_DEFERRED_ACTION_REQUIRE_NOEXCEPT -DSEASTAR_HAS_MEMBARRIER -DSEASTAR_HAVE_ASAN_FIBER_SUPPORT -DSEASTAR_HAVE_HWLOC -DSEASTAR_HAVE_NUMA -DSEASTAR_HAVE_URING -DSEASTAR_SCHEDULING_GROUPS_COUNT=16 -DSEASTAR_SHUFFLE_TASK_QUEUE -DSEASTAR_THREAD_STACK_GUARDS -DSEASTAR_TYPE_ERASE_MORE -Dseastar_EXPORTS -I/home/kefu/dev/seastar/include -I/home/kefu/dev/seastar/build/debug/gen/include -I/home/kefu/dev/seastar/src -g -std=gnu++20 -fPIC -U_FORTIFY_SOURCE -DSEASTAR_SSTRING -Wno-error=unused-result "-Wno-error=#warnings" -fstack-clash-protection -UNDEBUG -Wall -Werror -Wno-array-bounds -Wno-error=deprecated-declarations -gz -fsanitize=address -fsanitize=undefined -fno-sanitize=vptr -MD -MT CMakeFiles/seastar.dir/src/core/memory.cc.o -MF CMakeFiles/seastar.dir/src/core/memory.cc.o.d -o CMakeFiles/seastar.dir/src/core/memory.cc.o -c /home/kefu/dev/seastar/src/core/memory.cc
In file included from /home/kefu/dev/seastar/src/core/memory.cc:61:
/home/kefu/dev/seastar/include/seastar/util/log.hh:360:9: error: call to member function 'log' is ambiguous
        log(log_level::error, "{}", s);
        ^~~
/home/kefu/dev/seastar/include/seastar/util/log.hh:232:10: note: candidate function [with Args = <std::basic_string_view<char> &>]
    void log(log_level level, fmt::format_string<Args...> fmt, Args&&... args) noexcept {
         ^
/home/kefu/dev/seastar/include/seastar/util/log.hh:255:10: note: candidate function [with Args = <std::basic_string_view<char> &>]
    void log(log_level level, format_info fmt, Args&&... args) noexcept {
         ^
/home/kefu/dev/seastar/include/seastar/util/log.hh:368:9: error: call to member function 'log' is ambiguous
        log(log_level::warn, "{}", s);
        ^~~
```

Reply:
the idea was to templatize the other overload.  dropped in the latest revision.